lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -346,6 +368,184 @@ static LogicalResult verifyCircuitOp(CircuitOp circuit) {
+        if (!mod)
+          modFound = false;
+
+        if (!modFound) {
+          nla.emitOpError()
+              << " module:" << modName << " doesnot exist in the circuit";
+          return failure();
+        }
+        bool refFound = false;
+        // It can either be an inner_sym on a port.
+        for (auto sym : mod.getPortSymbols()) {
+          refFound = sym == innerRef.getName();
+          if (refFound)
+            break;
+        }
+        // Or it can be an inner_sym on any operation inside the module.
+        WalkResult opFound = WalkResult::interrupt();
+        if (!refFound)
+          opFound = mod.walk([&](Operation *op) {
+            auto attr = op->getAttrOfType<StringAttr>("inner_sym");
+            bool fallBackNameRefFound = false;
+            // fallBackNameRefFound, is is a hack, which looks for the case,
+            // when the inner_sym does not exist, and the symbol name is instead
+            // matched with the operation name. This case occurs when the
+            // nonlocal anchor is present on the field of a bundle. We don't
+            // want to add symbol to the entire bundle. LowerTypes is supposed
+            // to fix this case when the aggregate is lowered.
+            if (!attr)
+              fallBackNameRefFound =
+                  (op->getAttrOfType<StringAttr>("name") == innerRef.getName());
+            if (attr == innerRef.getName() || fallBackNameRefFound) {
+              AnnotationSet annos(op);
+              refFound = false;
+              for (auto anno : annos)
+                if (auto n = anno.getMember("circt.nonlocal")) {
+                  refFound = n.cast<FlatSymbolRefAttr>().getAttr() == nlaName;
+                  if (refFound)
+                    break;
+                }
+              if (!refFound) {
+                auto diag =
+                    nla.emitOpError()
+                    << " operation with symbol:" << innerRef
+                    << " does not contain a reference to the NonLocalAnchor";
+                diag.attachNote(op->getLoc()) << "the symbol was defined here";
+                return WalkResult::interrupt();
+              }
+              // The correct inner_sym found, now interrupt the module walk.
+              return WalkResult::interrupt();
+            }
+            return WalkResult::advance();
+          });
+        if (!opFound.wasInterrupted()) {
+          nla.emitOpError() << " operation with symbol:" << innerRef
+                            << " was not found in module:" << modName;
+          return failure();
+        }
+        if (!refFound)
+          return failure();
+
+      } else {
+        // This is the case when the nla is applied to a module.
+        if (expectedModuleName != ref.cast<FlatSymbolRefAttr>().getAttr()) {
+          nla.emitOpError()
+              << "expected module::" << expectedModuleName << " not found";
+        }
+      }
+    }
+  }
```

Comment:
High level: this logic gets pretty complicated pretty quickly and it would be excellent to try and simplify it.